Pattern Migration 2024: suggest nicer patterns

dianne opened the pull request:
This is based on #136817. The new commits start with removing a comment on the tests. The goal of the refactors in the following commits is to eventually [produce suggestions which remove reference patterns], fixing #136047 in a more thorough way.

Here's a simplified version of how it works, for reference (it's not actually split into discrete phases, but these are the conceptual steps): 
1. Try removing all reference patterns and binding modifiers; if this doesn't change any bindings, we can suggest the resultant pattern.
2. If a binding would be changed by this, we give it an explicit binding modifier. This requires adding explicit reference patterns for any dereferences above it (which may have previously been inserted implicitly by match ergonomics). This may change other bindings, in which case we'll need to suggest binding modifiers for them, which may require suggesting more reference patterns, etc.. We suggest whatever we end up with once we've fixed all such inconsistencies.
3. Simplify `&ref x` to `x` when it's convenient. This isn't quite complete; full simplification would require potentially rewriting `&ref x @ ...` to `x @ &...`, which would add a bit of complexity, though I can implement that too if desired.

Unfortunately, it ended up being more code than I'd hoped, so I understand if it's unappealing from a maintenance perspective. If nothing else, finalizing match ergonomics 2024 should open up significant simplifications, and I'm happy to help maintain it until then. There's a chance some common bookkeeping could be shared between other diagnostics/suggestions too; I have a couple in mind, but I haven't checked whether it'd be less complexity overall.

Relevant tracking issue: #131414

This doesn't aim to provide viable suggestions under the feature gates `ref_pat_eat_one_layer_2024` or `ref_pat_eat_one_layer_2024_structural`; some of the suggestions in the test output for those features are wrong and/or not as pretty as they could be. I've included FIXME comments providing more detail. Producing suggestions targeting the match ergonomics implemented by those features would actually be significantly simpler than what this PR does, but since it would require different[^1] logic, I'm inclined to save that for later. cc @Nadrieril, do any other feature gates come to mind that might require tweaks to this?

@rustbot label A-diagnostics A-patterns A-edition-2024 

[^1]: It would only require a small tweak to correct the suggestions in the `ref_pat_eat_one_layer_2024` and `ref_pat_eat_one_layer_2024_structural` tests. However, a different approach to the one in this PR would let us use simpler logic to give more local suggestions. I figure if we're adding logic specific to new match ergonomics, we may as well not (slightly) complicate this version of it further when we could be doing something better.

dianne committed:
remove comment on pattern migration tests about possible future improvements

The following commits are the possible future improvements.

dianne commented:
That should be possible, but it would require some logic. The nice thing is that tracking the default binding mode in the pattern the user wrote is already done by https://github.com/rust-lang/rust/pull/136475/commits/203d3109d8e96a6a4075205e836216d7cd281d5b, so I think it could be built on top of that. For minimal added logic, we could give up on removing binding modifiers if we suggest adding any reference patterns; we'd only suggest adding reference patterns if a `&`/`&mut` eats an inherited reference or we encounter a `ref x` binding under a `ref mut` dbm (or `ref mut x` under a `ref`, but that's a borrowck error anyway). For more precision beyond that, I think we'd have to start tracking ancestors and maybe also descendants, like this PR does: if we suggest adding a shared `&` pattern, the dbm in its descendants could change from `ref` to `ref mut`, so we'd at least need to know to bail when that happens. Accounting for that alone is maybe most of the complexity of this PR.

bors commented:
:umbrella: The latest upstream changes (presumably #136549) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne committed:
track patterns' dereferences in a forest

This will let us propagate changes to the default binding mode when
suggesting adding `&` patterns.

dianne committed:
separate suggestion building from default binding mode tracking

This lets us revise the suggestion on-the-fly as we discover which
reference patterns and binding modifiers are necessary vs. which may be
omitted.

dianne committed:
compute dbm in user's pattern from deref kind and mutability

A slight refactor: we'll always want to pass the mutability in to
`push_deref` to determine the default binding mode when a pattern is
removed, so this eliminates the redundancy of having both that and the
user's dbm as separate arguments.

dianne committed:
produce suggestions which remove reference patterns

This doesn't yet respect whether patterns are from macro expansions, so
it may try to suggest changes within an expansion (leading to a test
failure).

dianne committed:
simplify `&ref x` to `x`

dianne commented:
I've rebased on top of #136817 (which puts pattern migration in its own file) and pulled the commit structure apart a bit. Rebasing onto the other diagnostic improvements has made this a bit bigger in total than it was before, but hopefully each of the refactors should at least be a bit more digestible than they were initially.

bors commented:
:umbrella: The latest upstream changes (presumably #137235) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne committed:
don't suggest removals inside macro expansions

This fixes the regression in the previous commit.

dianne committed:
more tests for avoiding changes inside expansions

dianne committed:
add a failing test

dianne committed:
don't get trapped inside of expansions when trimming labels

dianne commented:
Rebased, plus a few changes:
- I've added a commit to make this avoid suggesting changes to patterns from macro expansions. I haven't added any feedback for cases where fixing the pattern would require messing with macro internals, but this at least allows suggestions in cases where that's not needed. I've also added some more tests for this.
- This is now based on #137269, which was necessary for annotating one of the tests I wrote for the other change.

wesleywiser commented:
Hi @Nadrieril, would you like us to re-roll for a different reviewer? 🙂

bors commented:
:umbrella: The latest upstream changes (presumably #139949) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nadrieril commented:
No need to reroll, I had semi-intentionally postponed this at the time. Will chat with @dianne about what we want to do.

dianne commented:
I'll close this for now; the gains this PR offers are pretty marginal compared to its complexity. It should be possible to do a simpler version of this, and that may be necessary for if Match Ergonomics 2024 stabilizes (the current migration logic doesn't produce nice suggestions there), but I'd prefer not to update the migration lint for it before the details there settle.
